File: unionfinds/2421-number-of-good-paths.go
package unionfinds

import "sort"

// https://leetcode.com/problems/number-of-good-paths/

// Approach: Union Find
// Time: O(nlogn)
// Space: O(n)
func numberOfGoodPaths(vals []int, edges [][]int) int {
	var n = len(vals)
	var parent = make([]int, n)
	var count = make([]int, n)
	for idx := range parent {
		parent[idx] = idx
		count[idx] = 1
	}
	var find = func(x int) int {
		for parent[x] != x {
			parent[x] = parent[parent[x]]
			x = parent[x]
		}
		return x
	}
	sort.Slice(edges, func(i, j int) bool {
		return max(vals[edges[i][0]], vals[edges[i][1]]) < max(vals[edges[j][0]], vals[edges[j][1]])
	})
	var px, py int
	var result = n
	for _, edge := range edges {
		px = find(edge[0])
		py = find(edge[1])
		if vals[px] < vals[py] {
			parent[px] = py
		} else if vals[py] < vals[px] {
			parent[py] = px
		} else {
			parent[py] = px
			result += count[px] * count[py]
			count[px] += count[py]
		}
	}
	return result
}

// // Approach: Brute-Force with Union Find (Time Limit Exceeded)
// // Time: O(?)
// // Space: O(?)
// func numberOfGoodPaths(vals []int, edges [][]int) int {
// 	var n = len(vals)
// 	var parent = make([]int, n)
// 	var nodes = make([]int, n)
// 	for node := range parent {
// 		parent[node] = node
// 		nodes[node] = node
// 	}
// 	var find = func(x int) int {
// 		for x != parent[x] {
// 			parent[x] = parent[parent[x]]
// 			x = parent[x]
// 		}
// 		return x
// 	}
// 	var union = func(x int, y int) {
// 		parent[find(x)] = find(y)
// 	}

// 	var adjList = make([][]int, n)
// 	var node1, node2 int
// 	for idx := range edges {
// 		node1, node2 = edges[idx][0], edges[idx][1]
// 		adjList[node1] = append(adjList[node1], node2)
// 		adjList[node2] = append(adjList[node2], node1)
// 	}
// 	sort.Slice(nodes, func(i, j int) bool { return vals[nodes[i]] < vals[nodes[j]] })
// 	var same = make(map[int]map[int]struct{})
// 	var result = n
// 	for _, node := range nodes {
// 		for _, neighbor := range adjList[node] {
// 			if vals[neighbor] <= vals[node] {
// 				union(neighbor, node)
// 			}
// 		}
// 		for twin := range same[vals[node]] {
// 			if find(twin) == find(node) {
// 				result++
// 			}
// 		}
// 		if same[vals[node]] == nil {
// 			same[vals[node]] = make(map[int]struct{})
// 		}
// 		same[vals[node]][node] = struct{}{}
// 	}
// 	return result
// }

File: unionfinds/2421-number-of-good-paths.py
from collections import defaultdict
from typing import List
import unittest

# https://leetcode.com/problems/number-of-good-paths/

# python3 -m unittest unionfinds/2421-number-of-good-paths.py

class Solution(unittest.TestCase):
    # # Approach: Brute-Force with Union Find (Time Limit Exceeded)
    # # Time: O(?)
    # # Space: O(?)
    # def numberOfGoodPaths(self, vals: List[int], edges: List[List[int]]) -> int:
    #     n = len(vals)
    #     parent = list(range(n))
    #     def find(x: int) -> int:
    #         while parent[x] != x:
    #             parent[x] = parent[parent[x]]
    #             x = parent[x]
    #         return x
    #     def union(x: int, y: int) -> None:
    #         px = find(x)
    #         py = find(y)
    #         parent[px] = py
    #     graph = defaultdict(list)
    #     for node1, node2 in edges:
    #         graph[node1].append(node2)
    #         graph[node2].append(node1)
    #     same = defaultdict(set)
    #     result = n
    #     for val, node in sorted([(vals[node],node) for node in range(len(vals))]):
    #         for neighbor in graph[node]:
    #             if vals[neighbor] <= val:
    #                 union(neighbor, node)
    #         for twin in same[vals[node]]:
    #             if find(twin) == find(node):
    #                 result += 1
    #         same[vals[node]].add(node)
    #     return result

    # Approach: Union Find
    # Time: O(nlogn)
    # Space: O(n)
    def numberOfGoodPaths(self, vals: List[int], edges: List[List[int]]) -> int:
        n = len(vals)
        parent = list(range(n))
        def find(x: int) -> int: # O(logn)
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x
        count = [1]*n
        result = n
        edges.sort(key=lambda edge: max(vals[edge[0]],vals[edge[1]]))
        for x, y in edges: # O(n)
            px, py = find(x), find(y)
            if vals[px] < vals[py]:
                parent[px] = py
            elif vals[px] > vals[py]:
                parent[py] = px
            else:
                parent[py] = px
                result += count[px] * count[py]
                count[px] += count[py]
        return result

    def test(self) -> None:
        for vals, edges, expected in [
            ([1], [], 1),
            ([2,1,1], [[0,1],[2,0]], 3),
            ([1,3,2,1,3], [[0,1],[0,2],[2,3],[2,4]], 6),
            ([1,1,2,2,3], [[0,1],[1,2],[2,3],[2,4]], 7),
        ]:
            output = self.numberOfGoodPaths(vals, edges)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
